AX: De-duplication in AXObjectCache::updateIsolatedTree is made redundant by de-duplication in AXIsolatedTree::queueNodeUpdate

twilco opened the pull request:
#### f5a9393bdeff7c89685de21aa9f2df392139cc07
<pre>
AX: De-duplication in AXObjectCache::updateIsolatedTree is made redundant by de-duplication in AXIsolatedTree::queueNodeUpdate
<a href="https://bugs.webkit.org/show_bug.cgi?id=297099">https://bugs.webkit.org/show_bug.cgi?id=297099</a>
<a href="https://rdar.apple.com/157812717">rdar://157812717</a>

Reviewed by Joshua Hoffman.

Beyond this de-duplication being redundant, this eliminates a strange issue where HashMap::get with an AXID not contained in the
now-deleted  `HashMap&lt;AXID, OptionSet&lt;Field&gt;&gt; updatedObjects` returned an OptionSet with a set bit, causing hundreds of
layout test failures. This HashMap::get usage has been around for quite a while, but somehow got broken by:

<a href="https://commits.webkit.org/298189@main">https://commits.webkit.org/298189@main</a>

Which seems extremely innocuous and unrelated to this code.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):

Canonical link: <a href="https://commits.webkit.org/298580@main">https://commits.webkit.org/298580@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/05dba80667b7e9c58ae98de3e83ef4a5393e6ee5

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/115936 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/35590 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/26126 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/121985 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/66469 "Built successfully") | [✅ 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/9dcc16aa-6fd1-416e-a841-82ac3b2ae8ca) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/117825 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/36279 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/44177 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/88068 "Passed tests") | [  ~~🧪 win-tests~~](https://ews-build.webkit.org/#/builders/60/builds/42667 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/92699e7b-e2c4-41dd-8b79-0fc7c00d5b52/2f8612bf-64ec-4509-add6-091b5fab9095) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/118884 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/28949 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/104039 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/68481 "Passed tests") | | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/28071 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/22147 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/65655 "Built successfully") | | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/98352 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/22281 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/125137 "Built successfully") | | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/42823 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/32153 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/96840 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/43189 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/100227 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/96609 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/24581 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/41873 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/19745 "Passed tests") | [  ~~🛠 playstation~~](https://ews-build.webkit.org/#/builders/134/builds/38782 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/42712 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/48301 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/42179 "Built successfully") | | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/45513 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/43886 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

hoffmanjoshua commented on Source/WebCore/accessibility/AXObjectCache.cpp:
```diff
@@ -4964,36 +4964,13 @@ void AXObjectCache::updateIsolatedTree(const Vector<std::pair<Ref<AccessibilityO
-            updatedFields.add(Field::DependentProperties);
-            updatedObjects.set(axObject->objectID(), updatedFields);
+        if (!idsWithUpdatedDependentProperties.contains(axObject->objectID())) {
+            idsWithUpdatedDependentProperties.add(axObject->objectID());
```
Nit: It's possible to make this one line, since `add` returns a `AddResult`, which has a bool operator to return if this is a newly added result:

```c++
if (idsWithUpdatedDependentProperties.add(axObject->objectID()))
    tree->updateDependentProperties(axObject.get());
```

But if you prefer the explicit check, I'm also good with that

twilco replied:
Fixed!

twilco committed:
AX: De-duplication in AXObjectCache::updateIsolatedTree is made redundant by de-duplication in AXIsolatedTree::queueNodeUpdate
https://bugs.webkit.org/show_bug.cgi?id=297099
rdar://157812717

Reviewed by Joshua Hoffman.

Beyond this de-duplication being redundant, this eliminates a strange issue where HashMap::get with an AXID not contained in the
now-deleted  `HashMap<AXID, OptionSet<Field>> updatedObjects` returned an OptionSet with a set bit, causing hundreds of
layout test failures. This HashMap::get usage has been around for quite a while, but somehow got broken by:

https://commits.webkit.org/298189@main

Which seems extremely innocuous and unrelated to this code.

* Source/WebCore/accessibility/AXObjectCache.cpp:
(WebCore::AXObjectCache::updateIsolatedTree):

Canonical link: https://commits.webkit.org/298580@main
